client/allocrunner/taskrunner/identity_hook.go

```diff
@@ -34,21 +47,42 @@ func (*identityHook) Name() string {
 func (h *identityHook) Prestart(ctx context.Context, req *interfaces.TaskPrestartRequest, resp *interfaces.TaskPrestartResponse) error {
 	h.lock.Lock()
 	defer h.lock.Unlock()
+	h.tokenPath = filepath.Join(req.TaskDir.SecretsDir, wiTokenFile)
 
-	token := h.tr.alloc.SignedIdentities[h.taskName]
-	if token != "" {
-		h.tr.setNomadToken(token)
-	}
-	return nil
+	return h.setToken()
 }
 
 func (h *identityHook) Update(_ context.Context, req *interfaces.TaskUpdateRequest, _ *interfaces.TaskUpdateResponse) error {
 	h.lock.Lock()
 	defer h.lock.Unlock()
 
+	return h.setToken()
+}
+
+// setToken adds the Nomad token to the task's environment and writes it to a
+// file if requested by the jobsepc.
+func (h *identityHook) setToken() error {
 	token := h.tr.alloc.SignedIdentities[h.taskName]
-	if token != "" {
-		h.tr.setNomadToken(token)
+	if token == "" {
+		return nil
 	}
+
+	h.tr.setNomadToken(token)
+
+	if id := h.tr.task.Identity; id != nil && id.File {
+		if err := h.writeToken(token); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// writeToken writes the given token to disk
+func (h *identityHook) writeToken(token string) error {
+	if err := ioutil.WriteFile(h.tokenPath, []byte(token), 0666); err != nil {
```

Comment:
```suggestion
	if err := ioutil.WriteFile(h.tokenPath, []byte(token), 0666); err != nil {
```

Does this file need to be world writable?

Reply:
Hm, this copied the Vault token file's behavior. My assumptions are:

1. This is necessary because we don't know what user the task will run as
2. This is safe because of directory permissions on the parent dir

If these aren't true, I'd love to fix them. If nothing else it's an awful smell.

Reply:
Hmm yeah, maybe there's something we could do if Task.User happens to be set?

But my real question is why does the file need to be _writable_? Would `0664` or even `0644` be sufficient?

Reply:
Just realized my 1-3 workflow above is broken by the fact that alloc updates will rewrite the token, so that use case is right out...

Maybe it shouldn't be world writable :thinking:

Reply:
The good: Checkout what I came up with in `helper/users.WriteFileFor`. It appears to degrade gracefully when `Task.User` is unset (even without the explicit guard I added) as well as properly set the mode and owner when running as `root` on Linux. All of this without any OS specific implementations so it should even work on any Unix.

The bad: I don't think there's anything bad here, but it definitely increases risk. I think that's tolerable for a new off-by-default feature.

The ugly: Check out the Dockerfile for our example job: https://github.com/docker-library/redis/blob/master/Dockerfile.template#L3-L4 It creates a user within the container and runs Redis as that. Setting `task.user = "nobody"` does still work. Redis runs as nobody within the container, and `nomad_secret` gets properly owned by nobody. Without setting `task.user` the fallback is used, so Redis, running as `redis`, is still able to access `nomad_secret`! So... it's ugly but it works?

I haven't tested it with Docker and user namespaces.

Reply:
@shoenig mind peeking at this and resolving the comment thread if it looks good to you?

Reply:
I added this comment about Windows
https://github.com/hashicorp/nomad/pull/15755/files#r1092065794